### ISSUE TITLE
[aes] Shall we enable SecAllowForcingMasks for ES?

### ISSUE THREAD

**Issue by vogelpi** [lowRISC/opentitan#14240]
At the moment, the masking in AES is always enabled on ASIC. Exclusively on FPGA, we allow disabling the masking using the `SecAllowForcingMasks` SV parameter in combination with the FORCE_ZERO_MASKS field in the main control register (on ASIC, the corresponding logic is simply not there). Having the option to disable the masking is useful to tune the SCA capture setup and evaluate the noise.

Recently, we have discussed in the SCA meeting and in particular with @johannheyszl and @m-temp that specifically for the positioning of the high-precision EM probes for our own evaluation of ES it would be important to have this feature also on ASIC. Otherwise, the positioning of the EM probe alone will be extremely time consuming and we won't have a way to evaluate how good the masking is compared to no masking. Therefore, I suggest the following:

1. Adapt the RTL to instead of forcing the masks to zero, forcing them to be constant by not advancing the masking PRNG if the Sec parameter and the control register are configured properly. From an SCA perspective, the effect of constant and all zero masks is the same - the masking is off. The advantage of forcing constant masks vs. zero masks is much lower area overhead. Disabling the Sec parameter for production will have a very low layout impact/could even be done with a metal fix. 
2. Move the FORCE_ZERO/CONSTANT_MASKS fields into the lockable CTRL_AUX_SHADOWED register. This has the advantage that the feature cannot be enabled anymore after some point.

It's also worth noting that in normal operation, a periodic reseed of the masking PRNG will happen after 1, 64 or 8k blocks. So even if someone managed to force constant masks, the PRNG state and thus the masks will be updated automatically at some point. For the lab evaluation, we would avoid this by setting the reseed interval to 8k, disabling the lockable KEY_TOUCH_FORCES_RESEED field in the CTRL_AUX_SHADOWED register, and by re-configuring the module after every 8k blocks (to avoid the reseed counter expiring).

Before changing the design, I would also like to get the opinion of @cdgori, @tjaychen and @moidx . WDYT?

**Comment by tjaychen** [comment]
@vogelpi just ack-ing that i see where you're coming from.
I have a couple of follow-up questions for @johannheyszl, but those probably can't be asked in the open, so I will get back to this hopefully early next week.

**Comment by johannheyszl** [comment]
@bilgiday to have a look too

**Comment by johannheyszl** [comment]
thanks @moidx @tjaychen @bilgiday for the pointers. Adding from the mail thread:
- There is a way to inject a known seed into the CSRNG (flag0) which is supplying the randomness to the AES core's PRNG for mask generation via the EDN. 
- This mechanism is protected against FI. (Short summary on this will be added by @bilgiday soon.)
- The current assumption is that based on such a known seed, we can compute all masks for SCA profiling with a minimal/no uncertainty.
- This make the design change described above unnecessary.
 
@msfschaffner @vogelpi do you see any potential issues in computing the masks?
If not we can close

**Comment by vogelpi** [comment]
Thanks @johannheyszl for your summary. Using the flag0 mechanism of CSRNG also sounds good to me. In the Security Working Group meeting we have been discussing the following challenges related to that:
1. We would need to ensure that no other EDN0 consumer is active during the AES capture. This is probably doable but as @tjaychen correctly suggested we need to prototype this before tapeout to be absolutely sure.
2. To be able to compute the masks, the lab will need to have access to the netlist. Because the actual masks depend on the EDN input to AES and some secret netlist constants. Sharing these constants might not be acceptable.

**Comment by johannheyszl** [comment]
cc @jadephilipoom for visibility

**Comment by vogelpi** [comment]
Looking into this right now. The following needs to be done:
- [ ] Change PRNG implementation to force instead of zeroing masks if the feature is enabled.
- [ ] FPGA SCA experiment with masking disabled and reseed timer disabled. The reseed timer is by default disabled on FPGA.
- [ ] Adapt CSR configuration layout
- [ ] FPGA SCA experiment with masking disabled and skipping reseeding by re-configuring AES before reseed timer expires.

**Comment by cdgori** [comment]
Apologies I did not see this one when originally filed - I was on holiday and missed that I was tagged in my efforts to dig out from post-holiday email, somehow.

I can see we looked at two paths and eventually re-settled on the one that @vogelpi wanted input on initially.

-I completely understand wanting to have this option for SCA setup characterization.  You still plan to move these fields into the lockable CTRL_AUX_SHADOWED, yes?

-The idea of not stepping the PRNG, rather than all-zero, is OK for me.  

-In general, I am worried about attacker use of this construct in some production situation.  How can we prevent this from happening (without our knowledge)?

Seems like the plan is:
-post-ES, do a metal-only fix to disable connection of these fields/ability to stop stepping the PRNG
Is there something else?

My utopian view is that I would like to fire an alert per-block if LC_STATE == PROD || PROD_END and CONSTANT_MASKS == 1, but I'm not sure how practical that is.  

(Basically to quarantine this to DEV parts only, but I suppose we need to discuss if SCA on PROD parts will be required for certification or other reasons.)

**Comment by vogelpi** [comment]
Update: After a lot of work, I've now managed to make CSRNG/EDN0 output the seed AES needs to produce zero masks. YIPPEE!

This uncovered a hardware bug in EDN (for details see #15402) and some issues in the DIF (see #15411). Next is to do the following:
- [ ] FPGA measurement on the CW310
- [ ] Working out a top-level tests, I think we should have one for this because the feature is very complex but also very important for ES.
- [ ] Prepare PRs

**Comment by vogelpi** [comment]
I've now done the FPGA measurement on the CW310 to verify that the EDN setup (see #15577) and mask forcing work as expected. Below figure shows some results.

![chart](https://user-images.githubusercontent.com/20307557/196557881-17a33a5a-dddc-48de-b4a9-c924e80a8b8a.png)

The y axis is leakage (higher means more leakage), the x axis is number of traces. The dotted lines are for the CW310 including EDN/CSRNG connection:
- red dots: masking on
- green dots: PRNG produces constant output but isn't loaded with the correct seed from EDN to produce all-zero output
- blue dots: PRNG produces constant output and is loaded with correct seed to produce all-zero output. The masking is off. In this scenario, we see many alerts firing due to repetition checks inside EDN consumers that also repetitively get the same seed.

For reference, the regular lines show the CW305 where we don't have EDN, i.e., the required seed to switch the masking off cannot be injected:
- red line: masking on
- green line: PRNG produces constant output
- blue line: PRNG produces constant, all-zero output. The masking is off. (just for reference, this is the old implementation that will not go into ES)

**Comment by johannheyszl** [comment]
Thank you Pirmin, this is great. IMO we are now in a position to achieve exactly what we aimed for thanks to your work!
Also thanks again for verifying. This was extremely important.

**Comment by vogelpi** [comment]
In the meantime, all the RTL and software changes have been merged. The last PR to merge before we can close this is https://github.com/lowRISC/opentitan/pull/15732 which add the DV component to the top-level test to actually probe into AES and ensure the masking is indeed off.

To answer to @cdgori 's questions (we discussed this or some of it in a meeting but I forgot to share the outcome here, sorry).
> -I completely understand wanting to have this option for SCA setup characterization. You still plan to move these fields into the lockable CTRL_AUX_SHADOWED, yes?

Yes, this is now done.

> -In general, I am worried about attacker use of this construct in some production situation. How can we prevent this from happening (without our knowledge)?

We discussed in the meeting that it is very hard to construct this because there are several hurdles to take (thanks to how the S-Boxes are masked):
- The force_masks bit inside CTRL_AUX_SHADOWED has to be changed, this requires bypassing the REGWEN protection and the shadow register. This keeps the PRNG output constant.
- The second share of the key needs to be all zero (I've forgot to mention this in the meeting). This means chances that someone is able to attack a sideload key are extremely low. 
- EDN/CSRNG have to be configured to produce the specific seed and not use any entropy.
- This will trigger many recoverable alerts in various places due to failing repetition checks in other EDN consumers. The alert handler needs to be bypassed for the system to continue operating. Alternatively, an adversary could reconfigure EDN/CSRNG after AES has loaded the required seed. This would avoid the repetition checks but the adversary now would have to avoid any reseeding operations of AES by reconfiguring before the internal, hardened block counter expires.

> Seems like the plan is:
> -post-ES, do a metal-only fix to disable connection of these fields/ability to stop stepping the PRNG
Is there something else?

Yes, and we've also been discussing to add a repetition check for EDN interface of AES. I've opened #15777 to not forget about this.

> My utopian view is that I would like to fire an alert per-block if LC_STATE == PROD || PROD_END and CONSTANT_MASKS == 1, but I'm not sure how practical that is.
>
> (Basically to quarantine this to DEV parts only, but I suppose we need to discuss if SCA on PROD parts will be required for certification or other reasons.)

Currently, LC_STATE is not available in AES and adding this is out of scope at the moment. But we should have this general discussion regarding SCA on PROD after M2.

**Comment by vogelpi** [comment]
Also the last PR for this issue has been merged. I am thus closing it. If something else comes up related to this, feel free to re-open or add a new issue.

**Comment by cdgori** [comment]
> We discussed in the meeting that it is very hard to construct this because there are several hurdles to take (thanks to how the S-Boxes are masked):
> 
> * The force_masks bit inside CTRL_AUX_SHADOWED has to be changed, this requires bypassing the REGWEN protection and the shadow register. This keeps the PRNG output constant.
> * The second share of the key needs to be all zero (I've forgot to mention this in the meeting). This means chances that someone is able to attack a sideload key are extremely low.
> * EDN/CSRNG have to be configured to produce the specific seed and not use any entropy.
> * This will trigger many recoverable alerts in various places due to failing repetition checks in other EDN consumers. The alert handler needs to be bypassed for the system to continue operating. Alternatively, an adversary could reconfigure EDN/CSRNG after AES has loaded the required seed. This would avoid the repetition checks but the adversary now would have to avoid any reseeding operations of AES by reconfiguring before the internal, hardened block counter expires.

This makes me feel much better.  Especially the point about the 2nd share needing to be all-zero!  But also all the other alerts that will fire.

> Yes, and we've also been discussing to add a repetition check for EDN interface of AES. I've opened #15777 to not forget about this.

Thank you!

> 
> > My utopian view is that I would like to fire an alert per-block if LC_STATE == PROD || PROD_END and CONSTANT_MASKS == 1, but I'm not sure how practical that is.
> > (Basically to quarantine this to DEV parts only, but I suppose we need to discuss if SCA on PROD parts will be required for certification or other reasons.)
> 
> Currently, LC_STATE is not available in AES and adding this is out of scope at the moment. But we should have this general discussion regarding SCA on PROD after M2.

Understood.

Thanks for the detailed write-up!
